DM-28859: Make the DBB endpoint buffer manager persist file sizes in the database

mxk62 opened the pull request:
Besides changes directly related to the feature request, I made some additional modifications. Most of them are cosmetic ones to make pylint complain a little bit less about the code. For example, I added few missing module docstrings and removed unnecessary `pass` statements.

However, I also replaced each component's `--validate` option with the corresponding, standalone command. The option was effectively turning each component into a validation tool as setting it on was making each component to terminate its execution *regardless* of the outcome of the validation. So I hope that capturing this functionality in a separate subcommand will provide a better "separation of concerns".

mxk62 committed:
Make finder store file sizes in the database

mxk62 committed:
Make `validate` a command

Each component of the manager allowed a user to validate the
configuration file by setting `--validate` option.  However, by design,
all components terminate their execution after validation is completed
regardless of its outcome.  As a result setting this option effectively
turns the `start` command into a validation tool.  For this reason I
removed `--validate` from `start`'s options and made it a separate
command.

mxk62 committed:
Remove unused variables

mxk62 committed:
Remove unnecessary pass statements

mxk62 committed:
Add missing module docstrings

mxk62 committed:
Fix import order

mxk62 committed:
Name argument consistently

Argument name differed between ABC and one of its concrete class.  Fixed
it to make pylint be a bit more happy about the code.

mxk62 committed:
Correct the name of the attribute in Macro class

Macro class was refering to the attribute of the Action class using
wrong name.  Fixed it.

mxk62 commented:
Besides changes directly related to the feature request, I made some additional modifications. Most of them are cosmetic ones to make `pylint` complain a little bit less about the code. For example, I added few missing module docstrings and removed unnecessary `pass` statements.

However, I also replaced each component's `--validate` option with the corresponding, standalone command. The `--validate` option was effectively turning each component into a validation tool as setting it on was making each component to terminate its execution *regardless* of the outcome of the validation. So I hope that capturing this functionality in a separate subcommand will provide a better "separation of concerns".

mxk62 committed:
Make backfill persist file sizes

I forgot to update backfill tool to make it persist files sizes in the
database.  Fixed it.

mxk62 committed:
Remove redundant section from the config file

The backfill tool had just one search-related option, exclude_list.
Having an entire configuraion section for it felt like an overkill.
Hence I removed it.

mxk62 committed:
Allow for ignoring failFast option

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -144,9 +143,9 @@ def run(self):
+                logger.debug("getting file attributes")
                 try:
-                    checksum = get_checksum(abspath)
+                    checksum, status = get_file_attributes(abspath)
```
To make this more future proof, really should return a container of all file attributes and pass that container (perhaps modified) to a function to save the file attributes.  Having each subcommand know exactly how many attributes there are   lessens the helpfulness of the get_file_attributes function.

mxk62 replied:
I don't think I quite understand what is your point here. It already returns a container, a tuple. You could pass it around without worrying about number of elements by writing, say,
```
attrs = get_file_attributes(abspath)
```
(can't modify it though without converting it to a list first). It just happens that I need to access its elements shortly after so I'm unpacking it "in-place". Do you want me to change the type of the container to something else?

MichelleGower replied:
Say next week we want to save another value per file (newattr1).  With the current code, everywhere get_file_attributes is called (which is more than this one line) the code would have to be changed to include newattr1 in the return tuple.   

If one has code similar in concept to:
```
mydict = get_file_attributes(abspath)
<code here that actually must use/modify certain mydict values>
add_file(self.session, mydict)
```
it doesn't matter here if mydict changes. The only changes in this case would go in the shared get_file_attributes and add_file functions.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/cli.py:
```diff
@@ -18,8 +18,10 @@
+"""
 import logging
+import click
 from .backfill.commands import backfill
```
PEP8 recommends: "You should put a blank line between each group of imports."

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:
```diff
@@ -82,3 +75,14 @@ def start(filename, dump, validate):
+    with open(filename) as f:
+        configuration = yaml.safe_load(f)
+    schema = yaml.safe_load(BACKFILL)
+    validate_config(configuration, schema)
```
Lines 85-87 are repeated with every validate_config call.  Is there a way to move them into the validate_config function?

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -208,17 +213,16 @@ def run(self):
+                    status = Status.INVALID
+                if message is not None:
+                    logger.debug("%s: message: '%s'", path, message)
+                    logger.debug("%s: status: '%s'", path, status.value)
```
Consider whether better to combine these two pieces of info into a single debugging message for easier searching.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/plugins.py:
```diff
@@ -175,7 +177,13 @@ def __init__(self, config):
+
+        # TODO: A temporary hack to make the plugin usable with LSST stack
+        #  prior to w_2021_05. Remove the if statement once a later version is
+        #  in use.
```
Better wording (i.e., the very first failFast doesn't guarantee we wouldn't roll back):  once the pre-failFast stack versions are old enough to never be used again.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/status.py:
```diff
@@ -18,6 +18,8 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""A list of available ingest statues.
```
typo:  statues -> statuses

MichelleGower commented on etc/backfill.yaml:
```diff
@@ -13,8 +13,7 @@ backfill:
     - <file source>
-  search:
-    exclude_list: null
+  exclude_list: null
```
Now only finder has search subsection with exclude_list in this subsection.  Consider whether worth having them at different level in finder than the other components.

MichelleGower replied:
Also the search section still exists in the config doc for backfill.

mxk62 replied:
I'll bring back the `search` section to the backfill config.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/finder.py:
```diff
@@ -167,6 +168,7 @@ def run(self):
                     relpath=os.path.relpath(dirname, start=self.storage),
                     filename=basename,
                     checksum=checksum,
+                    size_bytes=status.st_size,
```
Same comment as in backfill about better to have a function for the db operation to allow easier additions of more columns.

MichelleGower reviewed (approved):
A couple minor things and perhaps adding a db function to more future proof code.  Merge approved.

mxk62 committed:
Put a blank line between each group of imports

mxk62 committed:
Fix typos and language in comments

mxk62 committed:
Revert removing the config section from backfill

This reverts commit 7b3909be279e7bca90ef897335469b7121d4e9c7,
"Remove redundant section from the config file"

mxk62 committed:
Reduce duplicated code in validate commands

mxk62 committed:
Make logging messages more uniform

Backfill and Finder were logging slightly different messages doing the
same things.  Fixed that.

mxk62 committed:
Make get_file_attributes() return dataclass

Function get_file_attributes() was returning file attributes as a tuple.
I modified it to return a dataclass, FileInfo, instead to avoid the
necessity of memorizing the order in which it returns them.

mxk62 committed:
Add .gitignore file
